### ISSUE TITLE
Bucket Fretboard experiments by client_id

### ISSUE THREAD

**Issue by fbertsch** [mozilla-mobile/android-components#906]
Bucket users by client_id (ensures random distribution of clients, but consistent reproduction of experiments, esp. w.r.t. verifying branches). On our end we do crc32(client_id) % 100.

**Comment by st3fan** [comment]
From the Focus sprint planning I understand this is more an Q4 goal instead of an urgent fix needed for the current sprint? I think we should start the technical discussion sooner than later though.

**Comment by st3fan** [comment]
@fbertsch One concern here is what happens to clients in existing experiments. If we ship a change now, bucket numbers change and clients could suddenly switch from A to B. For example in the GeckoView experiment. Maybe that is fine, or maybe we need to roll this out at the right time when a reset doesn't matter?

Maybe @vesta0 can chime in too because I think that part is ultimately a product decision. The engineering side of this request is relatively simple.

**Comment by vesta0** [comment]
@st3fan I don't think we want to wait too long before making this change as it is blocking the A/A test so I'd like to understand exactly how this impacts the existing experiments. It sounds like it will impact the way existing experiment is set up and the ability to correctly interpret the results? In that case we would want to wait. What do you think @Sdaswani and @mmccorks ?

**Comment by mmccorks** [comment]
Is it possible to roll this out tomorrow with the 7.0 release? The experiment should be starting from scratch then.  Are we ok with people who have already gotten geckoview in beta getting webview in release? If we are, then that should work. If not, then we would have to wait to roll this out until the experiment in release is over, which would mean we have to wait to do the AA test until then. If we don't do this, will people on geckoview in beta stay on geckoview in release?

**Comment by vesta0** [comment]
thanks @mmccorks 

@st3fan is this something that can be rolled out tomorrow with the 7.0 release?

**Comment by Sdaswani** [comment]
looks like this was answered on slack, this will be blocked by the geckoview test.

**Comment by pocmo** [comment]
From the component side I think we need to provide two things:

* Telemetry: Expose the client id to consumers. Currently this is an internal detail and not shared outside of the library.
* Fretboard: Allow a consumer to provide an ID and only generate one if none is provided.

It would be up to the app to connect those two: Get the client ID from telemetry and pass it to Fretboard. With that it's also up to the app (team) when this happens.

The important thing is: If you change the ID used for bucketing then all users may end up in a different bucket. So the already ongoing 50:50 GeckoView experiment will be completely shuffled around again.

**Comment by Sdaswani** [comment]
hmm, so we are exposing something in one component so it can be used in another component, but the app has to set this up? seems like it breaks encapsulation / data hiding - can we provide a better interface to achieve this?

**Comment by pocmo** [comment]
> hmm, so we are exposing something in one component so it can be used in another component, but the app has to set this up? seems like it breaks encapsulation / data hiding - can we provide a better interface to achieve this?

As it stands fretboard and telemetry are two totally separated libraries and do not depend on each other. Fretboard already has a mechanism for the app to provide additional and alternative values (`ValuesProvider`) for experiment evaluation and with this change the "client id" is one of them.

**Comment by Sdaswani** [comment]
Fair enough @pocmo - I just expect @fbertsch wants all of our apps to do this, so it's potentially something we'll forget to do since it's a manual step.

**Comment by pocmo** [comment]
@Sdaswani That's a good point and the same is probably true for reporting the experiments in the various pings etc. - I'll file a separate issue to see if we can somehow make this whole setup easier for consumers of both libraries.

**Comment by Sdaswani** [comment]
great, thanks @pocmo
